File: example/c/sparse/heisenberg_crs_mpi.c
/*****************************************************************************
*
* Rokko: Integrated Interface for libraries of eigenvalue decomposition
*
* Copyright (C) 2012-2020 Rokko Developers https://github.com/t-sakashita/rokko
*    
* Distributed under the Boost Software License, Version 1.0. (See accompanying
* file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
*
*****************************************************************************/

#include <rokko/rokko.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(int argc, char *argv[]) {
  int provided;
  MPI_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &provided);
  int rank;
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  char* library_routine, *library, *routine;
  if (argc >= 2) {
    library_routine = rokko_parallel_sparse_ev_default_solver();
    if (argc >= 2) library_routine = argv[1];
  } else {
    library_routine = rokko_parallel_sparse_ev_default_solver();
  }
  rokko_split_solver_name(library_routine, &library, &routine);
  
  int L = (argc >= 3) ? atoi(argv[2]) : 10;
  int dim = 1 << L;
  int lattice_first[L], lattice_second[L];
  int l;
  for (l = 0; l < L; ++l) {
    lattice_first[l] = l;
    lattice_second[l] = (l+1) % L;
  }

  struct rokko_parallel_sparse_ev solver;
  rokko_parallel_sparse_ev_construct(&solver, library, argc, argv);
  struct rokko_mapping_1d map = rokko_parallel_sparse_ev_default_mapping(solver, dim, MPI_COMM_WORLD);
  struct rokko_distributed_crs_matrix mat;
  rokko_distributed_crs_matrix_construct(&mat, map, L+1);
  int row;
  int row_start = rokko_mapping_1d_start_row(map);
  int row_end = rokko_mapping_1d_end_row(map);
  int cols[dim];
  double values[dim];
  int count;
  double diag;
  int i, j, m1, m2, m3;
  for (row = row_start; row < row_end; ++row) {
    count = 0;
    diag = 0;
    for (l = 0; l < L; ++l) {
      i = lattice_first[l];
      j = lattice_second[l];
      m1 = 1 << i;
      m2 = 1 << j;
      m3 = m1 + m2;
      if (((row & m3) == m1) || ((row & m3) == m2)) {
        cols[count] = row^m3;
        values[count] = 0.5;
        ++count;
        diag += -0.25;
      } else {
        diag += 0.25;
      }
    }
    cols[count] = row;
    values[count] = diag;
    ++count;
    rokko_distributed_crs_matrix_insert(mat, row, count, cols, values);
  }
  rokko_distributed_crs_matrix_complete(mat);
  if (rank == 0) {
    printf("Eigenvalue decomposition of antiferromagnetic Heisenberg chain\n");
    printf("solver = %s\n", library);
    printf("L = %d\n", L);
    printf("dimension = %d\n", dim);
  }
  
  struct rokko_parameters params;
  rokko_parameters_construct(&params);
  // set some parameters
  if (routine[0] != '\0')  rokko_parameters_set_string(params, "routine", routine);
  rokko_parameters_set_int(params, "block_size", 5);
  rokko_parameters_set_int(params, "max_iters", 500);
  rokko_parameters_set_double(params, "conv_tol", 1.0e-12);
  rokko_parameters_set_int(params, "num_eigvals", 1);
  rokko_parallel_sparse_ev_diagonalize_distributed_crs_matrix(solver, mat, params);
  
  int num_conv = rokko_parallel_sparse_ev_num_conv(solver);
  printf("num_conv = %d\n", num_conv);
  if (num_conv == 0) MPI_Abort(MPI_COMM_WORLD, -1);
  int num_local_rows = rokko_distributed_crs_matrix_num_local_rows(mat);
  double eig_vec[num_local_rows];
  rokko_parallel_sparse_ev_eigenvector(solver, 0, eig_vec);

  if (rank == 0) {
    printf("number of converged eigenpairs = %d\n", num_conv);
    printf("smallest eigenvalues: ");
    for (i = 0; i < num_conv; ++i) printf("%30.20f", rokko_parallel_sparse_ev_eigenvalue(solver, i));
    printf("\n");
    printf("smallest eigenvector: ");
    for (j = 0; j < num_local_rows; ++j)
      printf("%30.20f ", eig_vec[j]);
    printf("\n");    
  }
  rokko_distributed_crs_matrix_destruct(&mat);
  rokko_parallel_sparse_ev_destruct(&solver);
  MPI_Finalize();
  return 0;
}

File: example/fortran/sparse/heisenberg_crs_mpi.f90
!/*****************************************************************************
!*
!* Rokko: Integrated Interface for libraries of eigenvalue decomposition
!*
!* Copyright (C) 2012-2013 by Tatsuya Sakashita <t-sakashita@issp.u-tokyo.ac.jp>,
!*                            Synge Todo <wistaria@comp-phys.org>,
!*                            Tsuyoshi Okubo <t-okubo@issp.u-tokyo.ac.jp>
!*
!* Distributed under the Boost Software License, Version 1.0. (See accompanying
!* file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
!*
!*****************************************************************************/

program heisenberg_crs_mpi
  use MPI
  use rokko
  implicit none
  integer :: provided, ierr, myrank, nprocs

  integer :: L, k
  integer :: dim
  integer, allocatable, dimension(:) :: lattice_first, lattice_second
  double precision :: diag
  integer :: i, j, m1, m2, m3, count
  integer :: row, start_row, end_row
  integer(c_int), allocatable, dimension(:) :: cols
  double precision, allocatable, dimension(:) :: values

  double precision :: eig_val
  double precision, allocatable, dimension(:) :: eig_vec

!  integer(4), allocatable, dimension(:) :: cols
!  double precision, allocatable, dimension(:) :: values

  type(rokko_parallel_sparse_ev) :: solver
  type(rokko_distributed_crs_matrix) :: mat
  character(len=100) :: solver_name, tmp_str
  integer :: arg_len, status
  type(rokko_parameters) :: params, params_out
  integer :: num_local_rows, num_conv

  call MPI_init_thread(MPI_THREAD_MULTIPLE, provided, ierr)
  call MPI_comm_rank(MPI_COMM_WORLD, myrank, ierr)
  call MPI_comm_size(MPI_COMM_WORLD, nprocs, ierr)

  if (command_argument_count() >= 1) then
     call get_command_argument(1, solver_name, arg_len, status)
  else
     call rokko_parallel_sparse_ev_default_solver(solver_name)
  endif

  if (command_argument_count() == 2) then  
     call get_command_argument(2, tmp_str, arg_len, status)
     read(tmp_str, *) L
  else
     L = 8
  endif
  
  dim = ishft(1,L)
  allocate( lattice_first(L) )
  allocate( lattice_second(L) )
  do k = 1, L
     lattice_first(k) = k - 1
     lattice_second(k) = mod(k, L)
  end do

  if (myrank == 0) then
     write(*,*) "solver name = ", trim(solver_name)
     write(*,*) "matrix dimension = ", dim
  endif

  call rokko_parallel_sparse_ev_construct(solver, solver_name)

  call rokko_distributed_crs_matrix_construct(mat, dim, dim, solver)

  start_row = rokko_distributed_crs_matrix_start_row(mat);
  end_row = rokko_distributed_crs_matrix_end_row(mat);

  allocate( cols(dim) )
  allocate( values(dim) )

  do row = start_row, end_row
     count = 0
     diag = 0
     do k = 1, L
        i = lattice_first(k)
        j = lattice_second(k)
        m1 = ishft(1, i)
        m2 = ishft(1, j)
        m3 = m1 + m2
        if ( (iand(row, m3) == m1) .or. (iand(row, m3) == m2) ) then
           count = count + 1
           cols(count) = ieor(row, m3)
           values(count) = 0.5
           diag = diag - 0.25
        else
           diag = diag + 0.25
        end if
     end do
     count = count + 1
     cols(count) = row
     values(count) = diag
     call rokko_distributed_crs_matrix_insert(mat, row, count, cols, values)
  end do
  call rokko_distributed_crs_matrix_complete(mat)
!  call rokko_distributed_crs_matrix_print(mat)

  call rokko_parameters_construct(params)
  call rokko_parallel_sparse_ev_diagonalize(solver, mat, params)
  
  num_conv = rokko_parallel_sparse_ev_num_conv(solver);
  num_conv = rokko_parallel_sparse_ev_num_conv(solver)
  if ((num_conv >= 1) .and. (myrank == 0)) then
     eig_val = rokko_parallel_sparse_ev_eigenvalue(solver, 0)
     print *, "eigval=", eig_val
  endif

  eig_val = rokko_parallel_sparse_ev_eigenvalue(solver, 0);
  num_local_rows = rokko_distributed_crs_matrix_num_local_rows(mat);
  print*, "num_local_rows=", num_local_rows
  allocate( eig_vec(num_local_rows) )
  call rokko_parallel_sparse_ev_eigenvector(solver, 0, eig_vec)

  if (myrank.eq.0) then
     print*, "number of converged eigenpairs=", num_conv
     print*, "Computed Eigenvalues = ", eig_val
     print*, "Computed Eigenvector = "
     print '(8f10.4)', eig_vec
  endif

  call rokko_distributed_crs_matrix_destruct(mat)
  call rokko_parallel_sparse_ev_destruct(solver)

  call MPI_finalize(ierr)
end program heisenberg_crs_mpi

